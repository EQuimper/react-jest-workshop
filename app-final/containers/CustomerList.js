/**
 * KENT: DON'T FORGET TO UPDATE THIS TO NOT
 * USE PROPS.STORE EVERYWHERE BEFORE YOUR
 * MIDWESTJS TALK :)
 */
import React, {PropTypes, Component} from 'react'
import store from '../store/Customers'

class CustomerList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      customers: props.store.getCustomers(),
    }
  }
  componentDidMount() {
    this.unsubscribe = this.props.store.subscribe(this.updateStateWithCustomers)
  }
  componentWillUnmount() {
    this.unsubscribe()
  }

  updateStateWithCustomers = () => {
    const customers = this.props.store.getCustomers()
    this.setState({customers})
  };

  render() {
    const {customers} = this.state
    if (customers.length === 0) {
      return <NoCustomers />
    } else {
      return <ListOfCustomers customers={customers} />
    }
  }
}

CustomerList.defaultProps = {store}

CustomerList.propTypes = {
  store: PropTypes.shape({
    getCustomers: PropTypes.func,
    subscribe: PropTypes.func,
  }),
}

function ListOfCustomers({customers}) {
  return (
    <div>
      Here is your list of customers!
      <ul>
        {customers.map((c, i) => <li key={i}>{c.name}</li>)}
      </ul>
    </div>
  )
}

ListOfCustomers.propTypes = {
  customers: PropTypes.array,
}

function NoCustomers() {
  return (
    <div>
      <div className="wrapper-that-needs-to-happen-for-some-reason">
        You have absolutely no customers. Better get to work!
      </div>
    </div>
  )
}

export default CustomerList
