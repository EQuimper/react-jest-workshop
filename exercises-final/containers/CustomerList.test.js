import React from 'react'
import renderer from 'react-test-renderer'
import getStoreStub from '../store/Customers.stub'
import CustomerList from './CustomerList'

test('should render no customers', () => {
  snapshotCustomerList()
})

test('should render customers', () => {
  const {store} = getStoreStub([{name: 'Bob'}, {name: 'Joanna'}])
  snapshotCustomerList({store})
})

test('should respond to store updates', () => {
  const {store, updateCustomers} = getStoreStub()
  const component = renderCustomerList({store})
  expect(component).toMatchSnapshot()
  updateCustomers([{name: 'Jill'}, {name: 'Fred'}])
  expect(component).toMatchSnapshot()
})

/**
 * Render the <CustomerList /> and snapshot it
 * @param {Object} props - the props to render with
 */
function snapshotCustomerList(props = {}) {
  const component = renderCustomerList(props)
  expect(component).toMatchSnapshot()
}

/**
 * Renders <CustomerList /> with the given props
 * @param {Object} props - the props to render with
 * @return {Object} the rendered component
 */
function renderCustomerList({store = getStoreStub().store}) {
  return renderer.create(<CustomerList store={store} />)
}
