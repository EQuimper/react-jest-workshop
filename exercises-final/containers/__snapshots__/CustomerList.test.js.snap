exports[`test should render customers 1`] = `
<div>
  <div>
    Here is your list of customers!
  </div>
  <ul>
    <li>
      Bob
    </li>
    <li>
      Joanna
    </li>
  </ul>
</div>
`;

exports[`test should render no customers 1`] = `
<div>
  You have no customers. Better get to work!
</div>
`;

exports[`test should respond to store updates 1`] = `
<CustomerList
  store={
    Object {
      "getCustomers": [Function getCustomers],
      "subscribe": [Function subscribe]
    }
  }>
  <function ListOfCustomers( /* istanbul ignore next */_ref2) {/* istanbul ignore next */var customers = _ref2.customers; /* istanbul ignore next */++__cov_lvVWst_SZwengrEOh37UL$X$OXw.f['6'];++__cov_lvVWst_SZwengrEOh37UL$X$OXw.s['13'];
    return (
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */null,
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */null, /* istanbul ignore next */'Here is your list of customers!'),
  
  
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'ul', /* istanbul ignore next */null,
          customers.map(function (c, i) {/* istanbul ignore next */++__cov_lvVWst_SZwengrEOh37UL$X$OXw.f['7'];++__cov_lvVWst_SZwengrEOh37UL$X$OXw.s['14'];return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'li', /* istanbul ignore next */{ key: i }, c.name));}))));
  
  
  
  }
    customers={
      Array [
        Object {
          "name": "Jill"
        },
        Object {
          "name": "Fred"
        }
      ]
    }>
    <div />
    <ul>
      <li />
      <li />
    </ul>
  </function ListOfCustomers( /* istanbul ignore next */_ref2) {/* istanbul ignore next */var customers = _ref2.customers; /* istanbul ignore next */++__cov_lvVWst_SZwengrEOh37UL$X$OXw.f['6'];++__cov_lvVWst_SZwengrEOh37UL$X$OXw.s['13'];
    return (
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */null,
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */null, /* istanbul ignore next */'Here is your list of customers!'),
  
  
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'ul', /* istanbul ignore next */null,
          customers.map(function (c, i) {/* istanbul ignore next */++__cov_lvVWst_SZwengrEOh37UL$X$OXw.f['7'];++__cov_lvVWst_SZwengrEOh37UL$X$OXw.s['14'];return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'li', /* istanbul ignore next */{ key: i }, c.name));}))));
  
  
  
  }>
</CustomerList>
`;
